passes/cmds/timeest.cc

```diff
@@ -0,0 +1,397 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2024  Martin Povišer <povik@cutebit.org>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/sigtools.h"
+#include "kernel/register.h"
+#include "kernel/cellaigs.h"
+#include "kernel/utils.h"
+#include "kernel/ff.h"
+#include "kernel/mem.h"
+
+#include <assert.h>
+#include <limits>
+
+USING_YOSYS_NAMESPACE
+template<> struct ::Yosys::hashlib::hash_ops<AigNode *> : Yosys::hashlib::hash_ptr_ops {};
+
+PRIVATE_NAMESPACE_BEGIN
+
+typedef long int arrivalint;
+const arrivalint INF_PAST = std::numeric_limits<arrivalint>::min();
+
+// each clock domain must have its own EstimateSta structure
+struct EstimateSta {
+	SigMap sigmap;
+	Module *m;
+	SigBit clk;
+
+	dict<std::pair<RTLIL::IdString, dict<RTLIL::IdString, RTLIL::Const>>, Aig> aigs;
+	dict<Cell *, Aig *> cell_aigs;
+
+	std::vector<std::pair<Cell *, SigBit>> launchers;
+	std::vector<std::pair<Cell *, SigBit>> samplers;
+	bool all_paths = false;
+	bool select = false;
+
+	void add_seq(Cell *cell, SigSpec launch, SigSpec sample)
+	{
+		sigmap.apply(launch);
+		sigmap.apply(sample);
+		launch.sort_and_unify();
+		sample.sort_and_unify();
+		for (auto bit : launch)
+			launchers.push_back(std::make_pair(cell, bit));
+		for (auto bit : sample)
+			samplers.push_back(std::make_pair(cell, bit));
+	}
+
+	int cell_type_factor(IdString type)
+	{
+		if (type.in(ID($gt), ID($ge), ID($lt), ID($le), ID($add), ID($sub),
+					ID($logic_not), ID($reduce_and), ID($reduce_or), ID($eq)))
+			return 1;
+		else
+			return 2;
```

Comment:
@aiju Some cells are discounted because they are implementable in dedicated carry chain logic and/or the model from `cellaigs.cc` is not balanced. This is very crude.